### ISSUE TITLE
Upgrade Rails and dependencies to current version

### ISSUE THREAD

**Issue by igrigorik** [igrigorik/vimgolf#173]
Discussion in https://github.com/igrigorik/vimgolf/issues/171.

- [ ] Follow Rails [upgrade guides](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-3-0-to-rails-3-1) to bring vimgolf to modernity.
- [ ] Upgrade all necessary dependencies - e.g. Mongoid, authentication layer, etc.

Non-goal: swap out data layer or change functionality. To start, I think we should migrate existing functionality a modern stack. With a modern stack in place, we can tackle other functional updates.

**Comment by zoso10** [comment]
dumping some resources for easier reference as we upgrade:

#### Release Announcements:
- [Rails 3.1.0](http://weblog.rubyonrails.org/2011/8/31/rails-3-1-0-has-been-released/)
- [Rails 3.2.0](http://weblog.rubyonrails.org/2012/1/20/rails-3-2-0-faster-dev-mode-routing-explain-queries-tagged-logger-store/)
- [Rails 4.0.0](http://weblog.rubyonrails.org/2013/6/25/Rails-4-0-final/)
- [Rails 4.1.0](http://weblog.rubyonrails.org/2014/4/8/Rails-4-1/)
- [Rails 4.2.0](http://weblog.rubyonrails.org/2014/12/19/Rails-4-2-final/)
- [Rails 5.0.0](http://weblog.rubyonrails.org/2016/6/30/Rails-5-0-final/)

#### Rails Upgrade Guides:
- [3.0 to 3.1](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-3-0-to-rails-3-1)
- [3.1 to 3.2](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-3-1-to-rails-3-2)
- [3.2 to 4.0](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-3-2-to-rails-4-0)
- [4.1 to 4.2](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-4-0-to-rails-4-1)
- [4.2 to 5.0](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-4-2-to-rails-5-0)

#### Extra Upgrade Resources:
- [Rails upgrade checklist](http://www.rails-upgrade-checklist.com/)
- [Railscasts 3.0 to 3.1](http://railscasts.com/episodes/282-upgrading-to-rails-3-1)
- [Railscasts 3.1 to 3.2](http://railscasts.com/episodes/318-upgrading-to-rails-3-2)
- [Railscasts 3.2 to 4.0](http://railscasts.com/episodes/415-upgrading-to-rails-4)
 - Other random [resources](https://gist.github.com/zoso10/6553a69a40023657fd3035405f4ad4fc) I've compiled

**Comment by zoso10** [comment]
after a quick look at the Gemfile. the current locked version `mongoid` is the only thing preventing us from rails 4.0 versions (it has a dependency of `activemodel (~> 3.0)`). so most version bumping should be at the code-level and not untangling dependencies.

in rails 3.1 they introduced the asset pipeline which mostly involved moving files around and changing extensions along with some config changes. I can go ahead and start diving into that unless anyone else wants to. I'll try to keep churn as small as possible.

**Comment by igrigorik** [comment]
👍 ... full speed ahead.

**Comment by hvenables** [comment]
Just to let you know I tried to upgrade to Rails 3.1 and adding the asset pipeline. I struggled to get sprockets 2.0.0 to work with the version of ruby that we are using. I found that i had to jump to rails 3.2 and sprockets up to 2.2.2 to get the asset pipeline working with our version of ruby. However this was just my experience...

**Comment by zoso10** [comment]
@hvenables thanks for the heads up! ya i've been finding similar issues so far. i was thinking it would be safer to leave out the asset pipeline for now and make sure everything is stable on the bumped rails versions, fix deprecations, then set up the asset pipeline. 

if you have any successful work, please feel free to open a pr!

**Comment by hvenables** [comment]
My branch got contaminated with having moved everything to asset pipeline, then i had to upgrade rails to 3.2.22.5 to make that work but that meant that mongoid had to be upgraded so things started to spiral.... It left me thinking if we can get to 3.1 without adding the asset pipeline that would be a start then we need to probably think about upgrading the database before we can move forward (or at least mongoid if later versions are compatible with our version of mongodb)

**Comment by zoso10** [comment]
dropping this here for easy access: [mongoid 5.0 upgrade](https://docs.mongodb.com/ruby-driver/master/tutorials/5.2.0/mongoid-upgrade/)

**Comment by igrigorik** [comment]
I believe we're good! We're up to Rails 5, all the specs are passing.

**Comment by igrigorik** [comment]
Oh, and... thank you, thank you, thank you everyone for all the hard work and help to make this happen! 👍 👍 👍
